test/type/self.swift

```diff
@@ -43,6 +43,20 @@ final class FinalMario : Mario {
     }
 }
 
+do {
+  final class GenericClass<T> {
+    func me() -> Self { self }
+
+    func invalidMe() -> Self {
+      GenericClass<Never>() // expected-error{{cannot convert return expression of type 'GenericClass<Never>' to return type 'Self'}}
+    }
```

Comment:
FYI: If I were to match the `U` and `T` (`U -> DynamicSelfType<T>`) in `simplifyRestrictedConstraintImpl` instead, the error here would become `cannot convert return expression of type 'C1<Int>' to return type 'C1<T>'`. Does not sound like an improvement to me though.